File: articles/calculus-for-systems-modeling/articles/flux-circulation-and-spatial-flow/fortran/flux_circulation_audit.f90
program flux_circulation_audit
  implicit none
  print '(A)', 'scenario segment_count approximate_flux approximate_circulation mean_tangential_alignment mean_normal_alignment field_description geometry_description warning'
  call audit(1.0d0, 16, 'coarse_circle')
  call audit(1.0d0, 64, 'medium_circle')
  call audit(1.0d0, 256, 'fine_circle')
contains
  subroutine audit(radius, segments, scenario)
    real(8), intent(in) :: radius
    integer, intent(in) :: segments
    character(len=*), intent(in) :: scenario
    integer :: i
    real(8) :: theta0,theta1,x0,y0,x1,y1,xm,ym,dx,dy,seglen,tx,ty,nx,ny,fx,fy
    real(8) :: flux_total,circulation_total,tangent_sum,normal_sum,pi
    character(len=32) :: warning
    pi = 4.0d0*atan(1.0d0)
    flux_total=0.0d0; circulation_total=0.0d0; tangent_sum=0.0d0; normal_sum=0.0d0
    do i=0,segments-1
      theta0 = 2.0d0*pi*i/segments
      theta1 = 2.0d0*pi*(i+1)/segments
      x0 = radius*cos(theta0); y0 = radius*sin(theta0)
      x1 = radius*cos(theta1); y1 = radius*sin(theta1)
      xm = 0.5d0*(x0+x1); ym = 0.5d0*(y0+y1)
      dx = x1-x0; dy = y1-y0
      seglen = sqrt(dx*dx+dy*dy)
      tx = dx/seglen; ty = dy/seglen
      nx = xm/radius; ny = ym/radius
      fx = -ym; fy = xm
      circulation_total = circulation_total + fx*dx + fy*dy
      flux_total = flux_total + (fx*nx + fy*ny)*seglen
      tangent_sum = tangent_sum + fx*tx + fy*ty
      normal_sum = normal_sum + fx*nx + fy*ny
    end do
    if (segments < 32) then
      warning = 'coarse_path'
    else
      warning = 'synthetic_flow'
    end if
    print '(A,1X,I8,4F14.6,1X,A,1X,A,1X,A)', trim(scenario), segments, flux_total, circulation_total, tangent_sum/segments, normal_sum/segments, 'field', 'circle', trim(warning)
  end subroutine
end program

File: articles/calculus-for-systems-modeling/articles/flux-circulation-and-spatial-flow/c/flux_circulation_audit.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

void vector_field(double x, double y, double* fx, double* fy){ *fx = -y; *fy = x; }
double dot2(double ax,double ay,double bx,double by){ return ax*bx + ay*by; }

void audit(double radius, int segments, const char* scenario){
  double flux_total = 0.0, circulation_total = 0.0, tangent_sum = 0.0, normal_sum = 0.0;
  for(int i=0; i<segments; i++){
    double theta0 = 2.0*M_PI*i/segments;
    double theta1 = 2.0*M_PI*(i+1)/segments;
    double x0 = radius*cos(theta0), y0 = radius*sin(theta0);
    double x1 = radius*cos(theta1), y1 = radius*sin(theta1);
    double xm = 0.5*(x0+x1), ym = 0.5*(y0+y1);
    double dx = x1-x0, dy = y1-y0;
    double segment_length = sqrt(dx*dx + dy*dy);
    double tx = dx/segment_length, ty = dy/segment_length;
    double nx = xm/radius, ny = ym/radius;
    double fx, fy;
    vector_field(xm, ym, &fx, &fy);
    circulation_total += dot2(fx, fy, dx, dy);
    flux_total += dot2(fx, fy, nx, ny) * segment_length;
    tangent_sum += dot2(fx, fy, tx, ty);
    normal_sum += dot2(fx, fy, nx, ny);
  }
  const char* warning = segments < 32 ? "Coarse path sampling; circulation and flux should be checked with more segments." : "Synthetic flow audit; document field meaning orientation units and boundary choice.";
  printf("%s,%d,%.12f,%.12f,%.12f,%.12f,%s,%s,%s\n", scenario, segments, flux_total, circulation_total, tangent_sum/segments, normal_sum/segments, "rotating field F=<-y,x>", "counterclockwise circle with radius 1", warning);
}

int main(void){
  printf("scenario,segment_count,approximate_flux,approximate_circulation,mean_tangential_alignment,mean_normal_alignment,field_description,geometry_description,warning\n");
  audit(1.0, 16, "coarse_circle");
  audit(1.0, 64, "medium_circle");
  audit(1.0, 256, "fine_circle");
  return EXIT_SUCCESS;
}
